### PR TITLE
Add a DiskArchver that implements a Cache TTL

### LINKED ISSUE #137: Cache TTL?
Another source of memory utilisation (see #136) is the no. of items we keep in the "global feed cache". Currently this is not limited or constrained in any way (_please don't start spamming your own feeds just to make this blow up on me! 🤣_)

See for example the following screenshot of my own timeline where I have 197 pages of twts I can page through. As if I'm going to do that! I've stopped caring after about page 10! 😂

It _might_ be a good idea to have a per-pod configurable cache ttl? What should a good default be if we decide we should do this? Should we just expunge items older than the ttl? Or move them to "cold storage" with a different storage (_e.g: a set of rotated files_)?

Screen Shot 2020-08-26 at 10 06 33

### PR DESCRIPTION
**Pull request by prologic**
##### Summary

ssia

##### Component Name

- area/backend

##### Test Plan

Tested locally

##### Additional Information

Closes #137

Creates a `./data/archive` structure like this:

Screen Shot 2020-08-27 at 13 55 11

### CODE REVIEW

**Comment by prologic** [comment]
I'm wondering whether we should also implement a `--max-cache-items` to limit the no. of cached twts per feed source, so whichever gets hits first is the point at which we start arching off to an `Archiver`? Hmmm 🤔

File: internal/archive.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+func (a *DiskArchiver) makePath(hash string) string {
+	chunks := Chunks(hash, 2)
+	for _, chunk := range chunks {
+		if len(chunk) != 2 {
+			chunk = fmt.Sprintf("0%s", chunk)
+		}
+	}
+
+	return filepath.Join(append([]string{a.path}, append(chunks, "twt.json")...)...)
```
Comment (dooven) [inline internal/archive.go:90]: im trying to understand what this method does 😅 Just wondering if you can show a screenshot of the directory structure with of an archived tweet

File: internal/cache.go [hunk 2]
```diff
@@ -182,15 +183,28 @@ func (cache Cache) FetchTwts(conf *Config, sources map[string]string) {
...
 
+				// Archive old twts
+				sort.Sort(sort.Reverse(old))
+				for _, twt := range old {
+					if archive.Has(twt.Hash()) {
+						// assume we have archived this twt and all older ones
+						break
+					}
+					if err := archive.Archive(twt); err != nil {
+						log.WithError(err).Errorf("error arching twt %s aborting", twt.Hash())
```
Comment (dooven) [inline internal/cache.go:201]: `error archiving`

File: internal/archive.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func (a *DiskArchiver) fileExists(fn string) bool {
+	if _, err := os.Stat(fn); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+		return false
```
Comment (dooven) [inline internal/archive.go:98]: this always return `false`. Should we drop the os.IsNotExist check?

File: internal/archive.go (hunk 1 above)
Comment (prologic) [reply internal/archive.go:90]: Yes! I will add it shortly to the desc above. Give me ~5mins for my local instance to do a feed cache cycle :D

File: internal/cache.go (hunk 2 above)
Comment (prologic) [reply internal/cache.go:201]: Fixed!

File: internal/archive.go (hunk 3 above)
Comment (prologic) [reply internal/archive.go:98]: Good point, we're not really caring about file system errors here. Fixed.

**Comment by prologic** [comment]
For comparison Git does something similar:

```#!sh
prologic@Jamess-iMac
Thu Aug 27 12:31:07
~/Projects/jointwt/twtxt/.git/objects
 (cache_ttl) 0
$ find . | wc -l
3974
```

Screen Shot 2020-08-27 at 12 31 04

**Comment by prologic** [comment]
I _might_ be able to make the `hashChunksSize` c constant and change it to a value of `4` and reduce the no. of directories and sub-directories. We _probably_ can't easily change this structure though once this is merge without writing tool (_so let's get this right!_)

**Comment by prologic** [comment]
I'm still not happy with the hash encoding... Still playing around with this. Going to hex encode the chunked hashes.

**Comment by prologic** [comment]
Okay I'm happy with the structure now:

Screen Shot 2020-08-27 at 13 55 11

**Comment by prologic** [comment]
Tested on `master` without and then again with the same data on `cache_ttl` branches to be sure we didn't break backwards compatibility:

With the configured defaults...

**Before:**

Screen Shot 2020-08-27 at 14 28 54

**After:**

Screen Shot 2020-08-27 at 14 35 56

**Comment by prologic** [comment]
I'm pretty happy with this now... The cache will be kept under control and by default shall not exceed:

- ~10 days worth of twts per feed source

or

- ~150 twts per feed source

Whichever comes first.

The `150` is based on the default 50 twts per age * 3

**Pull request merged**